moveit_kinematics/cached_ik_kinematics_plugin_description.xml

```diff
@@ -0,0 +1,38 @@
+<library path="lib/libmoveit_cached_ik_kinematics_plugin">
+  <class name="cached_ik_kinematics_plugin/CachedKDLKinematicsPlugin" type="cached_ik_kinematics_plugin::CachedIKKinematicsPlugin<kdl_kinematics_plugin::KDLKinematicsPlugin>" base_class_type="kinematics::KinematicsBase">
+    <description>
+      A kinematics plugin for persistently caching IK solutions computed with the KDL kinematics plugin.
+    </description>
+  </class>
+  <class name="cached_ik_kinematics_plugin/CachedSrvKinematicsPlugin" type="cached_ik_kinematics_plugin::CachedIKKinematicsPlugin<srv_kinematics_plugin::SrvKinematicsPlugin>" base_class_type="kinematics::KinematicsBase">
+    <description>
+      A kinematics plugin for persistently caching IK solutions computed with the Srv kinematics plugin.
+    </description>
+  </class>
+  <class name="cached_ik_kinematics_plugin/CachedTRACKinematicsPlugin" type="cached_ik_kinematics_plugin::CachedIKKinematicsPlugin<trac_ik_kinematics_plugin::TRAC_IKKinematicsPlugin>" base_class_type="kinematics::KinematicsBase">
+    <description>
+      A kinematics plugin for persistently caching IK solutions computed with the TRAC kinematics plugin.
+    </description>
+  </class>
+</library>
+<library path="lib/libmoveit_cached_ur3_kinematics_plugin">
+  <class name="cached_ik_kinematics_plugin/CachedUR3KinematicsPlugin" type="cached_ik_kinematics_plugin::CachedIKKinematicsPlugin<ur_kinematics::URKinematicsPlugin>" base_class_type="kinematics::KinematicsBase">
+    <description>
+      A kinematics plugin for persistently caching IK solutions computed with the UR3 kinematics plugin.
+    </description>
+  </class>
+</library>
+<library path="lib/libmoveit_cached_ur5_kinematics_plugin">
+  <class name="cached_ik_kinematics_plugin/CachedUR5KinematicsPlugin" type="cached_ik_kinematics_plugin::CachedIKKinematicsPlugin<ur_kinematics::URKinematicsPlugin>" base_class_type="kinematics::KinematicsBase">
+    <description>
+      A kinematics plugin for persistently caching IK solutions computed with the UR5 kinematics plugin.
+    </description>
+  </class>
+</library>
+<library path="lib/libmoveit_cached_ur10_kinematics_plugin">
+  <class name="cached_ik_kinematics_plugin/CachedUR10KinematicsPlugin" type="cached_ik_kinematics_plugin::CachedIKKinematicsPlugin<ur_kinematics::URKinematicsPlugin>" base_class_type="kinematics::KinematicsBase">
+    <description>
+      A kinematics plugin for persistently caching IK solutions computed with the UR10 kinematics plugin.
+    </description>
+  </class>
+</library>
```

Comment:
According to your explanation, the urX plugins are not supposed to be installed.
This xml is always installed though and registers the libraries with pluginlib.
That's of course a problem.

I believe the best way out is to move the ur5 code into a new package that does not even have to be part of this repository.
I don't think you can generate these xml files in the `CMakeLists.txt`...


Now that I think about it, we have the same problem with tracIK.
Do we want to make it a hard dependency in the package.xml to avoid it?
@mamoll The documentation indicates that tracIK does not gain anything from the caching, is that right?
Is it worth having a hard dependency in here or could/should we move the tracIK code into another (non-moveit) package?

Reply:
I can edit the CMakeLists.txt to generate the right XML file with some `configure_file(..)` magic. 

I made TracIK an optional dependency since I didn't want to increase the number of dependencies of moveit_kinematics. If you guys are ok with it, then we can make it a hard dependency. I don't have a strong preference one way or the other.

The TRAC IK *can* benefit from caching. For the Fetch planning for arm_and_torso move group can be slightly faster (performance can probably be further improved with some parameter tuning). I expect that for robots with more DOFs (humanoids, for instance), the benefits of caching can be quite substantial.

Reply:
> with some configure_file(..) magic

Yes, I thought about that too. The "magic" bit is the problem here.
You have to generate the file in-tree. This breaks static package analysis and it can break build servers if they set `src/` to read-only during the build (I know some people do that).

Unless you have more magic up your sleeves, I do not believe this solves the problem.
I'm sorry about this. I would really like to see this package available in the next release.
Is it that much overhead for you to have a separate repository for the specific plugins and only provide the template code + wrappers for moveit's internal plugins in `moveit_kinematics`?

Reply:
Hopefully, the last commit addresses your concern.

Reply:
I am very sorry, but no it does not.
If you build moveit in a devel space, ROS will not find the xml file and print
> ros.moveit_ros_planning.pluginlib.ClassLoader: Skipping XML Document "\<ws>/src/moveit_kinematics/cached_ik_kinematics_plugin_description.xml" which had no Root Element.  This likely means the XML is malformed or missing.

pluginlib (and I believe `rospack` too) only supports plugin xml files in the `src` folder in devel- and in the `share` folder in install-workspaces. Overall ROS barely supports optional compile-time dependencies up to now.

This is why I said
> You have to generate the file in-tree

earlier.

Do we want to live with that inherent problem?

Personally, I am perfectly happy to merge the part of this request that does not include ur and tracik,
and I would be happy to merge it if the cache-plugin would instantiate another kinematics_base-plugin it reads from a parameter, instead of using a C++ template that adds hard dependencies.
But these optional compile-time-dependencies break one way or another.

How do we proceed from here?